### PR TITLE
SSD512 Functional

### PR DESCRIPTION
**Pull request by ign-saurav**
### What's changed
Added TT implementation of SSD512 in `tt-metal` using TT-CNN architecture.

Input Resolution : 512x512x3
model: [SSD512](https://github.com/midasklr/SSD.Pytorch)

### PCC Scores:
| Head | Random Weights and input data |
|------|--------------|
| All Location outputs | > 0.99 | 
| All Confidence outputs | > 0.99 |

FPS: 66.5

Note
In the current demo implementation, the weights are random, so the object detection is not accurate in the inference.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by ign-saurav** [comment]
@sankarmanoj-tt, Could you please review the PR?

**Comment by tvardhineniTT** [comment]
@ign-saurav  
pcc and perf tests are failing with import issues . 

The test is importing : 
> from models.experimental.SSD512.reference.data.config import voc

But I observe there is no **/data** directory at all. 

Please fix imports and test.

**Comment by tvardhineniTT** [comment]
1. In readme, please update current performance, as PR shows 47 whereas Readme shows 49.9. 
2. In readme repository structure, test_e2e_performant.py is not mentioned. 
@ign-saurav

File: models/experimental/SSD512/demo/demo.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+from pathlib import Path
+import torch
+from loguru import logger
+import torch.nn.functional as F
+import ttnn
+import numpy as np
+import cv2
+from PIL import Image, ImageDraw, ImageFont
+
+sys.path.insert(0, str(Path(__file__).parent.parent.parent))
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/demo/demo.py]: What is this for? It appears to be a hack, that should be removed.

File: models/experimental/SSD512/demo/demo.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+
+    num_priors = loc_torch.shape[1] // 4
+    loc_torch = loc_torch.view(batch_size, num_priors, 4)
+    conf_torch = conf_torch.view(batch_size, num_priors, model.num_classes)
+
+    # Apply softmax to confidence
+    conf_torch = F.softmax(conf_torch, dim=-1)
+
+    # Convert priors to TTNN tensor
+    priors_ttnn = ttnn.from_torch(priors, device=device, dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT)
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/demo/demo.py]: Why do you convert loc from ttnn to torch and back to ttnn?

File: models/experimental/SSD512/demo/demo.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+        nonlocal device, ttnn_model
+
+        if device is not None:
+            ttnn.close_device(device)
+            import gc
+
+            gc.collect()
+
+        # Initialize TTNN device
+        if args.l1_small_size == 0:
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/demo/demo.py]: Why do you have the if condition? Why not just open with l1_small_size = args.l1_small_size?

File: models/experimental/SSD512/reference/box_utils.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
+# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
+# SPDX-License-Identifier: Apache-2.0
+
+# -*- coding: utf-8 -*-
+import torch
+
+
+def point_form(boxes):
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/reference/box_utils.py:8]: Are these utility functions used by any other models? If so, please don't replicate it.

File: models/experimental/SSD512/tt/layers/tt_extras_backbone.py [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+
+            is_1x1_conv = kernel_size == (1, 1) or (kernel_size[0] == 1 and kernel_size[1] == 1)
+
+            use_dram_slicing = (
+                not use_l1_for_this_layer
+                and (tensor_size_estimate > 1024 * 1024 or input_height > 64 or input_width > 64)
+                and not is_1x1_conv
+            )
+
+            if is_1x1_conv and stride == (1, 1) and padding == (0, 0):
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/tt/layers/tt_extras_backbone.py]: The conv2d op will internally call matmul for 1x1 kernel, stride 1 & padding 0. Why have you reimplemented this logic?

File: models/experimental/SSD512/tt/layers/tt_extras_backbone.py [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+                                        bias_torch.reshape((1, 1, 1, -1)),
+                                        dtype=ttnn.bfloat16,
+                                        layout=ttnn.ROW_MAJOR_LAYOUT,
+                                        device=None,
+                                    )
+                                else:
+                                    bias_host = None
+
+                                conv_config_for_weights = ttnn.Conv2dConfig(weights_dtype=ttnn.bfloat16)
+                                weight_prep = ttnn.prepare_conv_weights(
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/tt/layers/tt_extras_backbone.py]: Why have you used prepare_conv_weights? You can just keep the weights on host in the same format as Torch and call conv2d. It will work.

File: models/experimental/SSD512/tt/layers/tt_extras_backbone.py [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+                        weight_ttnn = ttnn.to_device(weight_ttnn, device)
+                        if bias_ttnn is not None:
+                            bias_ttnn = ttnn.to_device(bias_ttnn, device)
+                else:
+                    raise ValueError("Weight must be either a TTNN tensor or torch tensor")
+
+                estimated_memory_bytes = input_height * input_width * in_channels
+
+                if use_dram_slicing:
+                    slice_count = max(1, (batch_size * input_height * input_width) // (1024))
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/tt/layers/tt_extras_backbone.py]: You can set the slice_count to 0, and the op should automatically determine the most optimal number of slices.

File: models/experimental/SSD512/tests/pcc/test_extras_backbone.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+@pytest.mark.parametrize(
+    "pcc",
+    ((0.99),),
+)
+@pytest.mark.parametrize(
+    "size",
+    (512,),
+)
+@pytest.mark.parametrize("device_params", [{"l1_small_size": 24576}], indirect=True)
+def test_extras_backbone(device, pcc, size, reset_seeds):
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/tests/pcc/test_extras_backbone.py:22]: How many conv2d ops are present in the extras_backbone? Consider breaking the unit tests into smaller modules if the extras_backbone is large.

File: models/experimental/SSD512/tests/perf/test_e2e_performant.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        [dram_shard_height, padded_shard_width_channels],
+        ttnn.ShardOrientation.ROW_MAJOR,
+    )
+    dram_input_memory_config = ttnn.MemoryConfig(
+        ttnn.TensorMemoryLayout.HEIGHT_SHARDED, ttnn.BufferType.DRAM, dram_shard_spec
+    )
+
+    dram_output_memory_config = ttnn.MemoryConfig(
+        ttnn.TensorMemoryLayout.HEIGHT_SHARDED, ttnn.BufferType.DRAM, dram_shard_spec
+    )
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/tests/perf/test_e2e_performant.py]: Move this to a utility function. Also, why is DRAM sharded needed?

File: models/experimental/SSD512/tt/layers/detect.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+        boxes = torch.cat([center_pred, size_pred], dim=-1)  # [batch_size, num_priors, 4]
+
+        x1y1 = boxes[..., :2] - boxes[..., 2:] / 2
+        x2y2 = x1y1 + boxes[..., 2:]
+        boxes = torch.cat([x1y1, x2y2], dim=-1)  # [batch_size, num_priors, 4]
+
+        return boxes
+
+    def _nms(self, boxes: torch.Tensor, scores: torch.Tensor) -> torch.Tensor:
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/tt/layers/detect.py]: Don't reimplement nms. Use the same nms as the reference code

File: models/experimental/SSD512/tt/layers/tt_extras_backbone.py [hunk 11]
```diff
@@ -0,0 +1,575 @@
...
+    current_c = current_channels
+
+    for layer_idx, layer in enumerate(layers_with_weights):
+        if layer["type"] == "conv":
+            weight = layer["weight"]
+            bias = layer.get("bias", None)
+            config = layer["config"]
+            input_height = current_h
+            input_width = current_w
+            in_channels = current_c
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/tt/layers/tt_extras_backbone.py]: The entire conv2d code should be cleaned up. A lot of the logic implemented here is redundant or provided by the TT-CNN library.

File: models/experimental/SSD512/tt/tt_ssd.py [hunk 12]
```diff
@@ -0,0 +1,812 @@
...
+            config = layer["config"]
+
+            in_channels = layer["in_channels"]
+            out_channels = layer["out_channels"]
+            kernel_size = config["kernel_size"]
+            stride = config["stride"]
+            padding = config["padding"]
+            dilation = config["dilation"]
+            groups = config["groups"]
+
```
Comment (sankarmanoj-tt) [inline models/experimental/SSD512/tt/tt_ssd.py]: Same feedback as the extras backbone

**Review by sankarmanoj-tt — CHANGES_REQUESTED** [review]

**Comment by ign-saurav** [comment]
@sankarmanoj-tt @tvardhineniTT , The TT CNN implementation of SSD512 code has been refactored. The README and PR description have been updated with improved performance. Kindly review.

**Review by sankarmanoj-tt — APPROVED** [review]
